metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -0,0 +1,299 @@
+            considered_relevant_taxids.insert(full_parents_list[taxid]);
+        }
+
+        // Check if the current taxid is the root.
+        if (taxid == full_parents_list[taxid]) {
+            root_node = taxid;
+        }
+    }
+    if (num_taxid_failed) {
+        logger->warn("During the tax_tree_filepath {} parsing, {} taxids were not found out of {} total evaluations.",
+                     tax_tree_filepath, num_taxid_failed, relevant_taxids.size());
+    }
+
+    // Construct the output tree.
+    for (const TaxId &taxid : relevant_taxids) {
+        if (taxid == root_node) {
+            continue;
+        }
+        (*tree)[full_parents_list[taxid]].push_back(taxid);
+    }
+}
+
+void TaxonomyClsAnno::dfs_statistics(const TaxId node,
+                                     const ChildrenList &tree,
+                                     std::vector<TaxId> *tree_linearization) {
+    node_to_linearization_idx[node] = tree_linearization->size();
+    tree_linearization->push_back(node);
+    uint32_t depth = 0;
+
+    auto it = tree.find(node);
+    if (it != tree.end()) {
+        for (const TaxId &child : it->second) {
+            dfs_statistics(child, tree, tree_linearization);
+            tree_linearization->push_back(node);
+            if (node_depth[child] > depth) {
+                depth = node_depth[child];
+            }
+        }
+    }
+    node_depth[node] = depth + 1;
+}
+
+void TaxonomyClsAnno::rmq_preprocessing(const std::vector<TaxId> &tree_linearization) {
+    uint32_t num_rmq_rows = log2(tree_linearization.size()) + 1;
+
+    rmq_data.resize(num_rmq_rows);
+    for (uint32_t i = 0; i < num_rmq_rows; ++i) {
+        rmq_data[i].resize(tree_linearization.size());
+    }
+
+    // Copy tree_linearization to rmq[0].
+    for (uint32_t i = 0; i < tree_linearization.size(); ++i) {
+        rmq_data[0][i] = tree_linearization[i];
+    }
+
+    // Delta represents the size of the RMQ's sliding window (always a power of 2).
+    uint32_t delta = 1;
+    for (uint32_t row = 1; row < num_rmq_rows; ++row) {
+        for (uint32_t i = 0; i + delta < tree_linearization.size(); ++i) {
+            // rmq_data[row][i] covers an interval of size delta=2^row and returns the node with the maximal depth
+            // among positions [i, i+2^row-1] in the linearization.
+            // According to 'this->dfs_statistics()':
+            //     node_depth[leaf] = 1 and node_depth[root] = maximum distance to a leaf.
+
+            if (node_depth[rmq_data[row - 1][i]] >
+                node_depth[rmq_data[row - 1][i + delta]]) {
+                rmq_data[row][i] = rmq_data[row - 1][i];
+            } else {
+                rmq_data[row][i] = rmq_data[row - 1][i + delta];
+            }
+        }
+        delta *= 2;
+    }
+}
+
+std::vector<TaxId> TaxonomyClsAnno::get_lca_taxids_for_seq(const std::string_view &sequence, bool reversed) const {
+    cerr << "Assign class not implemented reversed = " << reversed << "\n";
```

Comment:
`std::cerr`

Why does it not require `std::`?
Do we have `using namespace std;` written somewhere?

Reply:
why not use the logger?

Reply:
added everything to the `throw std::runtime_error` line. I had to use all the parameters to not get compilation errors.

There is no `using namespace std;` in `metagraph/src`, but there are a few in the external libraries. Maybe it is taken from there.